### ISSUE TITLE
UseShellExecute on macOS passes Arguments to `open` instead of specified `filename`

### ISSUE THREAD

**Issue by SteveL-MSFT** [dotnet/corefx#30718]
Using PSCore6:

```powershell
$si = [System.Diagnostics.ProcessStartInfo]::new()
$si.UseShellExecute = $true
$si.FileName = (Get-Command pwsh).Source
$si.Arguments = "-c 1+1;read-host"
[System.Diagnostics.Process]::Start($si)
```

Expected:

Start new instance of PowerShell outputting answer of `2` and waiting for `Enter`, then exit.

Actual:

Help for `open` command is displayed since `-c` is not valid.

When using `open`, arguments to the executable are passed as `--args <arguments>`

The above script works as expected on Windows.

**Comment by karelz** [comment]
@SteveL-MSFT how does 'open' get into play here? Is that something .NET Core invokes when UseShellExecute=true?
Is the bug that on Mac we should prepend args with --args?

**Comment by SteveL-MSFT** [comment]
@karelz Yes, .NET Core invokes `open` on macOS and `xdg-open` on Linux when UseShellExecute=true.  On macOS, the `Arguments` property isn't passed correctly, so in the above example, .Net Core actually creates a process with arguments

> open pwsh -c 1+1;read-host

`open` treats `-c 1+1;read-host` as arguments to itself, not `pwsh`, so it fails (since `-c` is not a valid parameter to `open`) and outputs help text.  Looking at the help text for `open`, the correct format that should be used is:

> open pwsh --args -c 1:1;read-host

Probably something similar for the Linux code path, but I only tested this on macOS.

**Comment by karelz** [comment]
That sounds fairly straightforward. @wtgodbe what do you think?
@SteveL-MSFT is it blocking PowerShell scenarios?

**Comment by wtgodbe** [comment]
Seems reasonable. Looking into code paths now.

To start with, looks like this is the start of the bad path: 
https://github.com/dotnet/corefx/blob/4e27e0ad415dda38c2c2add84f0315a6f56406b3/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L298-L302

**Comment by wtgodbe** [comment]
Looks like we just need to modify https://github.com/dotnet/corefx/blob/4e27e0ad415dda38c2c2add84f0315a6f56406b3/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L398-L426 to append `--args` into the arg string in the case that `UseShellExecute=true`, since in that case, we always set the filename to the path to `open`:

https://github.com/dotnet/corefx/blob/4e27e0ad415dda38c2c2add84f0315a6f56406b3/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L298-L302

**Comment by SteveL-MSFT** [comment]
@karelz this is not blocking PowerShell Core.  Since enabling UseShellExecute on non-Windows is new capability, to workaround this, if the user passed Arguments, we simply set UseShellExecute to false.  not ideal, but sufficient for now.  The main scenario we wanted to enable is supporting `start-process https://bing.com` to work on non-Windows as it currently works on Windows.

**Comment by wtgodbe** [comment]
@karelz @SteveL-MSFT do either of you guys know enough about powershell to know how to best test a fix? I have a fix prepared, but I don't know how to get powershell to target my local build of CoreFx.

**Comment by SteveL-MSFT** [comment]
@wtgodbe I believe you can just overwrite the corefx assembly in $PSHOME

**Comment by wtgodbe** [comment]
Actually in testing without my change, I get a different result than you did. My output is:

> 2
> /bin/sh: read-host: command not found

However, running `read-host` on its own works as expected.

**Comment by wtgodbe** [comment]
I'm also having trouble testing with a local version of CoreFx - after copying in CoreFx libraries, I'm getting errors like:

>Unhandled Exception: System.TypeInitializationException: The type initializer for 'System.Management.Automation.PSVersionInfo' threw an exception. ---> System.IO.FileNotFoundException: /usr/local/microsoft/powershell/6.0.2/System.Management.Automation.dll
   at System.Diagnostics.FileVersionInfo.GetVersionInfo(String fileName)
   at System.Management.Automation.PSVersionInfo..cctor()
   --- End of inner exception stack trace ---
   at System.Management.Automation.PSVersionInfo.get_GitCommitId()
   at System.Management.Automation.Tracing.SysLogProvider.Log(PSEventId eventId, PSChannel channel, PSTask task, PSOpcode opcode, PSLevel level, PSKeyword keyword, Object[] args)
   at System.Management.Automation.Tracing.PSSysLogProvider.WriteEvent(PSEventId id, PSChannel channel, PSOpcode opcode, PSLevel level, PSTask task, PSKeyword keyword, Object[] args)
   at System.Management.Automation.Tracing.PSEtwLog.LogOperationalInformation(PSEventId id, PSOpcode opcode, PSTask task, PSKeyword keyword, Object[] args)
   at Microsoft.PowerShell.UnmanagedPSEntry.Start(String consoleFilePath, String[] args, Int32 argc)
   at Microsoft.PowerShell.ManagedPSEntry.Main(String[] args)

**Comment by SteveL-MSFT** [comment]
@wtgodbe that exception is complaining it can't find SMA.dll at that location.  Did you only overwrite the single assembly?

**Comment by wtgodbe** [comment]
I did, in that case I get:

>The shell cannot be started. A failure occurred during initialization:
Could not load file or assembly 'System.ComponentModel.Primitives, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

If I follow the chain & copy over the other dependencies (`System.ComponentModel.Primitives` and `System.Runtime`), I get:

> FailFast: The type initializer for 'Microsoft.PowerShell.ApplicationInsightsTelemetry' threw an exception.

   at System.Environment.FailFast(System.String, System.Exception)
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.PowerShell.UnmanagedPSEntry.Start(System.String, System.String[], Int32)
   at Microsoft.PowerShell.ManagedPSEntry.Main(System.String[])

**Comment by SteveL-MSFT** [comment]
The alternative is to allow passing args, but document that on non-Windows, the args get passed the default app (on Windows, it does get passed to the executable at least in the `pwsh` test case).  The downside to supporting that is the end user needs to know exactly what the default app is.  For example, if they open a URL, then the command line args passed to Chrome may not work with Firefox.  In which case, it would be better for them to invoke Chrome or Firefox directly.

**Comment by tmds** [comment]
The original issue reported here should be fixed by https://github.com/dotnet/corefx/pull/33052.

**Comment by danmosemsft** [comment]
@tmds @SteveL-MSFT can we close this then? If there is more to do then another issue might be clearest

**Comment by SteveL-MSFT** [comment]
If this is fixed in 3.0, we can close this.  Not sure when we can move PowerShell Core 6 to 3.0.

**Comment by danmosemsft** [comment]
fixed then.
